components/api/src/main/java/com/hotels/styx/api/FullHttpRequest.java

```diff
@@ -549,38 +570,91 @@ public Builder method(HttpMethod method) {
+         * <p>
+         * Note that this requires decoding the current header value before re-encoding, so it is most efficient to
+         * add all new cookies in one call to the method rather than spreading them out.
          *
-         * @param name  cookie name
-         * @param value cookie value
-         * @return {@code this}
+         * @param cookies new cookies
+         * @return this builder
          */
-        public Builder addCookie(String name, String value) {
-            return addCookie(HttpCookie.cookie(name, value));
+        public Builder addCookies(RequestCookie... cookies) {
+            return addCookies(asList(cookies));
         }
 
         /**
-         * Removes a cookie if present (removes its Set-Cookie header).
+         * Adds cookies into this request by overwriting the value of the "Cookie" header.
+         * <p>
+         * Note that this requires decoding the current header value before re-encoding, so it is most efficient to
+         * add all new cookies in one call to the method rather than spreading them out.
          *
-         * @param name name of the cookie
-         * @return {@code this}
+         * @param cookies new cookies
+         * @return this builder
          */
-        public Builder removeCookie(String name) {
-           findCookie(cookies, name)
-                    .ifPresent(cookies::remove);
+        public Builder addCookies(Collection<RequestCookie> cookies) {
+            Set<RequestCookie> currentCookies = decode(headers.get(COOKIE));
 
-            return this;
+            List<RequestCookie> combinedCookies = new ArrayList<>(currentCookies.size() + cookies.size());
+            combinedCookies.addAll(cookies);
+            combinedCookies.addAll(currentCookies);
+            return cookies(combinedCookies);
+        }
+
+        /**
+         * Removes all cookies matching one of the supplied names by overwriting the value of the "Cookie" header.
+         *
+         * @param names cookie names
+         * @return this builder
+         */
+        public Builder removeCookies(String... names) {
+            return removeCookies(asList(names));
+        }
+
+        /**
+         * Removes all cookies matching one of the supplied names by overwriting the value of the "Cookie" header.
+         *
+         * @param names cookie names
+         * @return this builder
+         */
+        public Builder removeCookies(Collection<String> names) {
+            return removeCookiesIf(toSet(names)::contains);
+        }
+
+        private Builder removeCookiesIf(Predicate<String> removeIfName) {
+            Predicate<RequestCookie> keepIf = cookie -> !removeIfName.test(cookie.name());
+
+            List<RequestCookie> newCookies = decode(headers.get(COOKIE)).stream()
+                    .filter(keepIf)
+                    .collect(toList());
+
+            return cookies(newCookies);
+        }
```

Comment:
This is perhaps bit contrived example, but perhaps worth looking into @kvosper:

```
    @Test
    public void removesCookies2() {
        HttpRequest r1 = HttpRequest.get("/")
                .addCookies(requestCookie("x", "x1"))
                .removeCookies("x")
                .build();

        assertThat(r1.cookie("x"), is(Optional.empty()));
    }
```

This test would fail as follows:

```
java.lang.AssertionError: 
Expected: is <Optional.empty>
     but: was <Optional[x=x1]>
```

I'll find this counter-intuitive.